From ef483019d025d9a7d73279ad99ec41c259aa56ca Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Mon, 11 Aug 2025 12:41:50 +0800
Subject: [PATCH 069/105] mtk: mt76: mt7996: deny addba request when delete BA
 is not finish

MT7996 needs to delete RRO session when receiving FW command.
Deny addba request until delete BA finish to prevent removing
incorrect RRO session.

Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>
---
 mt7996/init.c   | 15 +++++++++++++++
 mt7996/main.c   |  4 ++++
 mt7996/mcu.c    |  1 +
 mt7996/mt7996.h | 12 ++++++++++++
 4 files changed, 32 insertions(+)

diff --git a/mt7996/init.c b/mt7996/init.c
index dfefbe9a0..c620df448 100644
--- a/mt7996/init.c
+++ b/mt7996/init.c
@@ -1314,6 +1314,21 @@ reset:
 		}
 reset_session:
 		mt7996_mcu_wed_rro_reset_sessions(dev, e->id);
+
+		if (is_mt7996(&dev->mt76)) {
+			struct ieee80211_sta *sta;
+			struct mt7996_sta *msta;
+			struct mt76_wcid *wcid;
+
+			rcu_read_lock();
+			wcid = rcu_dereference(dev->mt76.wcid[e->wcid]);
+			sta = wcid_to_sta(wcid);
+			if (sta) {
+				msta = (struct mt7996_sta *)sta->drv_priv;
+				msta->stop_rx_ba_in_progress = false;
+			}
+			rcu_read_unlock();
+		}
 out:
 		kfree(e);
 	}
diff --git a/mt7996/main.c b/mt7996/main.c
index 7f963f866..84c250774 100644
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -1720,6 +1720,10 @@ mt7996_ampdu_action(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 
 	switch (params->action) {
 	case IEEE80211_AMPDU_RX_START:
+		if (msta->stop_rx_ba_in_progress) {
+			ret = -EAGAIN;
+			break;
+		}
 		/* Since packets belonging to the same TID can be split over
 		 * multiple links, store the AMPDU state for reordering in the
 		 * primary link
diff --git a/mt7996/mcu.c b/mt7996/mcu.c
index ffe52447c..154d8bbac 100644
--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -1383,6 +1383,7 @@ mt7996_mcu_wed_rro_event(struct mt7996_dev *dev, struct sk_buff *skb)
 				break;
 
 			session->id = le16_to_cpu(e->session_id);
+			session->wcid = le16_to_cpu(e->mld_id);
 
 			spin_lock_bh(&dev->wed_rro.lock);
 			list_add_tail(&session->list, &dev->wed_rro.poll_list);
diff --git a/mt7996/mt7996.h b/mt7996/mt7996.h
index 0b90d254f..4f68e7664 100644
--- a/mt7996/mt7996.h
+++ b/mt7996/mt7996.h
@@ -438,6 +438,8 @@ struct mt7996_sta {
 	u8 seclink_id;
 
 	struct mt7996_vif *vif;
+
+	bool stop_rx_ba_in_progress;
 };
 
 struct mt7996_vif_link {
@@ -525,6 +527,7 @@ struct mt7996_wed_rro_addr {
 struct mt7996_wed_rro_session_id {
 	struct list_head list;
 	u16 id;
+	u16 wcid;
 };
 
 struct mt7996_msdu_page {
@@ -595,6 +598,15 @@ struct mt7996_wmm_params {
 	int acm;
 };
 
+#define MT7996_MAX_RRO_RRS_RING 4
+
+struct mt7996_rro_cidx_didx_emi {
+	struct {
+		u16 idx;
+		u16 rsv;
+	} ring[MT7996_MAX_RRO_RRS_RING];
+};
+
 #ifdef CONFIG_MTK_VENDOR
 #define MT7996_AIR_MONITOR_MAX_ENTRY	16
 #define MT7996_AIR_MONITOR_MAX_GROUP	(MT7996_AIR_MONITOR_MAX_ENTRY >> 1)
-- 
2.45.2

